cmd/opdoc/opdoc.go

```diff
@@ -0,0 +1,243 @@
+// Copyright (C) 2019 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/algorand/go-algorand/config"
+	"github.com/algorand/go-algorand/data/transactions/logic"
+	"github.com/algorand/go-algorand/protocol"
+)
+
+func opGroupMarkdownTable(og *logic.OpGroup, out io.Writer) {
+	fmt.Fprint(out, `| Op | Description |
+| --- | --- |
+`)
+	// TODO: sort by logic.OpSpecs[].Opcode
+	for _, opname := range og.Ops {
+		fmt.Fprintf(out, "| `%s` | %s |\n", markdownTableEscape(opname), markdownTableEscape(logic.OpDoc(opname)))
+	}
+}
+
+func markdownTableEscape(x string) string {
+	return strings.ReplaceAll(x, "|", "\\|")
+}
+
+func enumTableMarkdown(out io.Writer, names []string) {
+	fmt.Fprintf(out, "| Index | Name |\n")
+	fmt.Fprintf(out, "| --- | --- |\n")
+	for i, name := range names {
+		fmt.Fprintf(out, "| %d | %s |\n", i, markdownTableEscape(name))
+	}
+	out.Write([]byte("\n"))
+}
+
+func fieldTableMarkdown(out io.Writer, names []string, types []logic.StackType) {
+	fmt.Fprintf(out, "| Index | Name | Type |\n")
+	fmt.Fprintf(out, "| --- | --- | --- |\n")
+	for i, name := range names {
+		gfType := types[i]
+		fmt.Fprintf(out, "| %d | %s | %s |\n", i, markdownTableEscape(name), markdownTableEscape(gfType.String()))
+	}
+	out.Write([]byte("\n"))
+}
+
+func transactionFieldsMarkdown(out io.Writer) {
+	fmt.Fprintf(out, "\n`txn` Fields:\n\n")
+	fieldTableMarkdown(out, logic.TxnFieldNames, logic.TxnFieldTypes)
+}
+
+func globalFieldsMarkdown(out io.Writer) {
+	fmt.Fprintf(out, "\n`global` Fields:\n\n")
+	fieldTableMarkdown(out, logic.GlobalFieldNames, logic.GlobalFieldTypes)
+}
+
+func opToMarkdown(out io.Writer, op *logic.OpSpec) (err error) {
+
+	opextra := logic.OpImmediateNote(op.Name)
+	cost := logic.OpCost(op.Name)
+	fmt.Fprintf(out, "\n## %s\n\n- Opcode: 0x%02x %s\n", op.Name, op.Opcode, opextra)
+	if op.Args == nil {
+		fmt.Fprintf(out, "- Pops: _None_\n")
+	} else if len(op.Args) == 1 {
+		fmt.Fprintf(out, "- Pops: *... stack*, %s\n", op.Args[0].String())
+	} else if len(op.Args) == 2 {
+		fmt.Fprintf(out, "- Pops: *... stack*, {%s A}, {%s B}\n", op.Args[0].String(), op.Args[1].String())
```

Comment:
The general case below seems to handle `len(op.Args)==2` in the same way that this special case handles it; any reason for having this extra code?

Reply:
simplified